sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -726,8 +726,9 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
 
     writeLong(array.length)
     writeLongArray(writeBuffer, array, array.length)
-    val used = ((cursor - Platform.LONG_ARRAY_OFFSET) / 8).toInt
-    writeLong(used)
+    val cursorFlag = cursor - Platform.LONG_ARRAY_OFFSET
+    writeLong(cursorFlag)
+    val used = (cursorFlag / 8).toInt
```

Comment:
Is this said, that when `(cursor - Platform.LONG_ARRAY_OFFSET) / 8` is over the range of Int, we will have overflow? But later you still do `toInt` and use the value?

Reply:
losing data when serializing LongHashedRelation in executor, can you see [this picture](http://oi67.tinypic.com/2z5pzs7.jpg)? In executor, the cursor is 0.

Reply:
Can you post the image in this PR? The web site you refer contains too many ads.

Reply:
![image](https://issues.apache.org/jira/secure/attachment/12932027/Spark%20LongHashedRelation%20serialization.svg)